py/dml/ctree.py

```diff
@@ -1396,7 +1396,7 @@ def make(cls, site, lh, rh):
         if ((lhtype.is_arith and rhtype.is_arith)
             or (isinstance(lhtype, (TPtr, TArray))
                 and isinstance(rhtype, (TPtr, TArray))
-                and compatible_types(lhtype.base, rhtype.base))):
+                and compatible_types_fuzzy(lhtype.base, rhtype.base))):
```

Comment:
This looks somewhat scary: `Compare` is a widely 1.4 thing, and still uses the ill-defined fuzzy cmp. If we want a single compat feature for lenient type checking, then we should consider bundling fuzzy cmp with it?

If we don't have the time to port all forgiveness corners in the short term, then one option is to start strict, i.e., let everyone say:
```
def compatible_types(lh, rh):
  if lenient in enabled_compat:
      return compatible_types_fuzzy(lh, rh)
  ...
```
If we later see that this is too strict, then it's safe to relax relevant corner later.

Reply:
> This looks somewhat scary: Compare is a widely 1.4 thing, and still uses the ill-defined fuzzy cmp.

Like I've said earlier:

> The primary point of this PR is to introduce a mechanism for strict type equality (which we can leverage for e.g. idempotent externs), and secondarily to convert most aspects of DMLC to leverage that mechanism. The remaining parts that still uses `cmp_fuzzy` constitute the most difficult cases, where some kind of type unification or conversion are needed (such as the ternary operator and comparison operators.)

I even note in the documentation of `fuzzy_cmp` and `compatible_types_fuzzy` that all uses of them should be considered HACKs. In particular, we can't use `compatible_types` here for `Compare` as that would mean you couldn't compare a void pointer with any other pointer.

I rather not introduce even more mechanisms _as part of this PR_ to cover compares and the ternary operator unless you insist upon it.

Reply:
ok maybe actually the `Compare` case is moot? Maybe the following is good enough:
```
    ...
    and (lhtype.base.void or rhtype.base.void
            or compatible_types(safe_realtype_unconst(lhtype.base),
                                safe_realtype_unconst(rhtype.base)))):
```

I think I'll need to look at the C standard to be sure.

Reply:
C standard says for `==` and `!=`:
> One of the following shall hold:
> * both operands have arithmetic type;
> * both operands are pointers to qualified or unqualified versions of compatible types;
> * one operand is a pointer to an object or incomplete type and the other is a pointer to a
qualified or unqualified version of `void`; or
> * one operand is a pointer and the other is a null pointer constant.
Case 3 and 4 are separated to forbid comparing equality between a function pointer and a void pointer. I suggest we don't bother with that.

C standard says for `<`, `>`, `<=`, `>=`:
> One of the following shall hold:
> * both operands have real type;
> * both operands are pointers to qualified or unqualified versions of compatible object types; or
> * both operands are pointers to qualified or unqualified versions of compatible
incomplete types

This excludes function pointers entirely, but also, to my surprise, even excludes void pointers. I don't know why, honestly. Well if C forbids it we have no reason to permit it.

Regardless, these rules are simple enough to not need `compatible_types_fuzzy`. Now lemme check the ternary case...

Reply:
> I rather not introduce even more mechanisms as part of this PR to cover compares and the ternary operator unless you insist upon it.

Fixing up remaining fuzzy cmp will introduce a different kind of type checking lenience. I can see some options:
* Make this compat feature unstable/provisional-ish, so we can stricten it in a later PR (requires that we add some base support for unstable compats)
* Plan to add unfuzzing as a separate compat feature later. Preparation should include renaming this feature to something more specific, to make it easy to distinguish from the unfuzzing compat feature.
* Blindly unfuzz all typechecks when lenient is disabled, and loosen the rules in a later PR (i.e. what I suggested earlier).

Reply:
Bleh.

> One of the following shall hold for the second and third operands:
> * both operands have arithmetic type;
> * both operands have the same structure or union type;
> * both operands have `void` type;
> * both operands are pointers to qualified or unqualified versions of compatible types;
> * one operand is a pointer and the other is a null pointer constant; or
> * one operand is a pointer to an object or incomplete type and the other is a pointer to a
qualified or unqualified version of `void`

I'm not even sure what to make of this. In particular, I don't know if there's any difference between point 2 (when taking into account point 1) and "both operands are qualified or unqualified versions of compatible object types"

Reply:
I don't think ternary `?:` needs to be that hard: We don't need the full complexity of "common type" in C; it should suffice if numeric are handled as by other binary operands, and otherwise require operands to be of compatible types (possibly modulo shallow_unconst).

Reply:
> it should suffice if numeric are handled as by other binary operands, and otherwise require operands to be of compatible types (possibly modulo shallow_unconst).
Nope, we need void-pointer compatibility. Though that's not a particularly hard exception to add either.

Reply:
Really the hardest part would be, as you say, common type-ing. In particular, const-unification.

Reply:
wait we already have code for that, it just has to be improved a slight bit. Yeah. Ok. This actually does seem plausible! We can get rid of the non-gated `compatible_type_fuzzy`s!

Reply:
What I mean is that we can start with a restrictive baseline: completely forbid `foo_t *` : `const foo_t *` to start with, see what breaks, and conservatively soften the ban to polish the roughest corners.

We'll immediately notice that `char *:const char *` is needed, for `cond?variable:"literal"`, and for that it should suffice to say that if we have two pointers, then compatibility of target types (modulo const) is sufficient for acceptance, where const in either operand takes precedence. This may forbid some corners that should be accepted, but if that's rare then it's procrastinatable.

Reply:
Now gotten rid of all ungated usages of `cmp_fuzzy`.